### PR TITLE
Makefile fixes

### PR DESCRIPTION
**Pull request by nmeyerhans**
### Summary

Makefile improvements to address #977 and #978 

### Implementation details

General makefile cleanup to provide better handling of dependencies.

### Testing

Multiple runs of `make clean`, `make get-deps`, etc. Confirm that subsequent runs don't do anything, e.g. 

```
ulap:amazon-ecs-agent‹makefile-fixes›$ make get-deps    
if [ -d amazon-ecs-cni-plugins/plugins ]; then \
        cd amazon-ecs-cni-plugins && git fetch && git checkout master ; \
else \
        git clone https://github.com/aws/amazon-ecs-cni-plugins.git --branch master ; \
fi
Cloning into 'amazon-ecs-cni-plugins'...
remote: Counting objects: 1867, done.
remote: Total 1867 (delta 0), reused 0 (delta 0), pack-reused 1867
Receiving objects: 100% (1867/1867), 2.91 MiB | 1.88 MiB/s, done.
Resolving deltas: 100% (808/808), done.
Checking connectivity... done.
mkdir -p amazon-ecs-cni-plugins/bin/plugins
touch .cni-sources-stamp
go get github.com/tools/godep
go get golang.org/x/tools/cmd/cover
go get github.com/golang/mock/mockgen
go get golang.org/x/tools/cmd/goimports
touch .get-deps-stamp
ulap:amazon-ecs-agent‹makefile-fixes›$ make get-deps  
make: Nothing to be done for 'get-deps'.
```

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [ ] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [ ] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: There are no specific tests for the makefile.

### Description for the changelog
This change doesn't introduce any behavior changes into the agent and can be left out of the changelog.

### Licensing

This contribution is under the terms of the Apache 2.0 License: yes

### CODE REVIEW

**Comment by samuelkarp** [comment]
So I think I'm not really convinced that I want all of these dependencies grouped together into the same target.  Things like the `go get` commands really shouldn't ever be automatically run (they're not necessary for building and only for development) and *should* be run on-demand.  On the other hand, the `get-cni-sources` target *should* be run automatically and for builds, and shouldn't be unnecessarily re-run.

What do you think about making these just separate targets entirely?  You can have target dependencies on `get-cni-sources` instead of `get-deps` and remove the stamping for `get-deps`.

**Comment by nmeyerhans** [comment]
@samuelkarp Yes, I think I agree with you. We should download the CNI sources as a dependency for our agent targets.

I wonder also if we should avoid deleting the CNI sources in the `clean` target. That target can delete any actual build but leave the sources in place. We could add an additional `distclean` target that goes further to reset the source tree to a pristine state, including blowing away the CNI sources.

**Comment by samuelkarp** [comment]
`get-cni-sources` seems like a poorly implemented copy of something like a git submodule, so it might just be better to use a real submodule.

If I remember correctly, we did this because the CNI plugin sources were moving very quickly as they were under active development.  Things are mostly stable now, so I think the benefit of using a system that's designed for this (embedding one git repo inside another) and tightly controls the version of the thing that's there outweighs the annoyance of changing the revision.  One of the Go vendoring tools might also work here, but we'll have to figure out how to get it to work in a nice way because this isn't a typical use-case for those tools.

**Comment by nmeyerhans** [comment]
Added another commit that converts the `amazon-ecs-cni-plugins` to a submodule.

File: Makefile [hunk 1]
```diff
@@ -97,83 +97,88 @@ test-registry: netkitten volumes-test squid awscli image-cleanup-test-images flu
...
 	rm -rf out/*
-	rm -rf $(shell pwd)/amazon-ecs-cni-plugins
+	$(MAKE) -C $(ECS_CNI_REPOSITORY_SRC_DIR) clean
 	rm -rf agent/Godeps/_workspace/pkg/
 	-$(MAKE) -C misc/netkitten $(MFLAGS) clean
 	-$(MAKE) -C misc/volumes-test $(MFLAGS) clean
 	-$(MAKE) -C misc/gremlin $(MFLAGS) clean
 	-$(MAKE) -C misc/testnnp $(MFLAGS) clean
 	-$(MAKE) -C misc/image-cleanup-test-images $(MFLAGS) clean
+	-rm -f .get-deps-stamp .cni-sources-stamp
```
Comment (samuelkarp) [inline Makefile]: `.cni-sources-stamp` doesn't exist anywhere.

Comment (nmeyerhans) [reply Makefile]: Fixed, thanks.

**Pull request merged**